### PR TITLE
Generate report with phys2bids outputs

### LINKED ISSUE #131: Generate a report with all the outputs
As discussed in our last meeting, it would be very interesting to have some kind of report that contains all the outputs phys2bids currently provides. This feature would unify our current three outputs into one "good looking" report.

## Context / Motivation

This feature would provide an easy way of looking at the output of phys2bids. It would also be useful for the tutorials on readthedocs.

## Possible Implementation

I guess an HTML file makes sense for this purpose.

### PR DESCRIPTION
**Pull request by eurunuela**
Closes #131 

## Proposed Changes

  - Adds the functionality to generate a report with phys2bids outputs on every call to phys2bids.

### CODE REVIEW

**Comment by eurunuela** [comment]
At the moment, and until we close #242, the report only shows the contents of the log file. Once we have the participant files and we restructure the output into folders, the report will show the folder structure and the plots generated by phys2bids.

@smoia let me know if the Brainhack and Enhancement labels should be removed.

**Comment by 62442katieb** [comment]
Hi!  👋 I'd like to help with report generating and I'm curious about what you all would ultimately like the report to look like/contain.
> At the moment, and until we close #242, the report only shows the contents of the log file. Once we have the participant files and we restructure the output into folders, the report will show the folder structure and the plots generated by phys2bids.
Based on this comment, report would contain folder structure + plots? Dynamic plots?! ( 😄)

Let me know! (I know it's late, though, no rush)

**Comment by eurunuela** [comment]
Hey Katie! 

Ideally it should contain the folder structure generated by physbids, as well as the channel plot (if requested by the user) and the trigger plot. That’s my opinion though.

@smoia anything else you think we should cover?

**Comment by smoia** [comment]
_... Did anyone say "dynamic"?_

If you want to add a bit of bokeh magic, please do so!
The channel plot, if interactive, could become a quality check!

**Comment by eurunuela** [comment]
Hey @62442katieb ! Any update on this?

**Comment by 62442katieb** [comment]
> Hey @62442katieb ! Any update on this?

Hey @eurunuela ! I've been out of office recently, but will get to this by the end of the week and have an update for you then!

**Comment by 62442katieb** [comment]
@eurunuela, I have an update! I managed to fix the issue with the arrangement of the log and plot on the report html. I'm making some minor formatting tweaks and I'll push my changes.

**Comment by eurunuela** [comment]
Awesome! I can't wait to see how gorgeous the report looks 🙈 thank you 62442katieb

**Comment by eurunuela** [comment]
Do we want a report per run or do we want one report were runs are concatenated; i.e. treated as channels of the same run?

What do you think @physiopy/phys2bids ?

**Comment by eurunuela** [comment]
> Do we want a report per run or do we want one report were runs are concatenated; i.e. treated as channels of the same run?
> 
> What do you think @physiopy/phys2bids ?

I'd like to bring this discussion back @physiopy/all

**Comment by tsalo** [comment]
@eurunuela can we make it so the integration test retains the report to make it easier to review? Like we do in `tedana`?

**Comment by eurunuela** [comment]
> @eurunuela can we make it so the integration test retains the report to make it easier to review? Like we do in `tedana`?

Of course! I'll have to do that during the weekend though.

**Comment by eurunuela** [comment]
@tsalo I can't figure out why the reports are not being saved. Any idea?

File: phys2bids/tests/test_integration.py
Comment (tsalo) [inline phys2bids/tests/test_integration.py:77]: I think these are copying to a file named `reporting`, not copying both files to a new `reporting` folder. It's possible that just adding a trailing `/` will fix it.

```suggestion
    shutil.copy(join(test_path, 'phys2bids_report.html'), '/tmp/src/phys2bids/reporting/')
    shutil.copy(join(test_path, 'phys2bids_report_log.html'), '/tmp/src/phys2bids/reporting/')
```

Comment (eurunuela) [reply phys2bids/tests/test_integration.py:77]: Okay, let's try it ou!

**Comment by eurunuela** [comment]
Okay, I was missing the deeper phys2bids directory on my paths. The reports are now saved but the plots won't load for some reason. Any idea @tsalo ?

**Comment by tsalo** [comment]
Are the figures stored in the actual HTML file or does the HTML just have references to files with the figures? If it's the latter, then maybe the actual figure files aren't being uploaded as artifacts.

**Comment by eurunuela** [comment]
The figures are actually not figures but some Bokeh-generated JS code that populates the figures with a JSON dictionary that is on the HTML file.

**Comment by tsalo** [comment]
I downloaded some of the files locally and it ended up rendering the images after a long delay. I don't think I downloaded _enough_ of the files for the report to look 100%, and it freezes every time I move a figure, but it's enough to say that this might just be a time issue.

**Comment by eurunuela** [comment]
We could just show just part of the time series to fix the loading issue, but I'd expect users to want to see the entire thing.

**Comment by vinferrer** [comment]
Once this is done, either add it to the documentation or create an issue for it

**Comment by eurunuela** [comment]
Well, it was definitely an issue with the number of points I was trying to plot. I've reduced it to 50 samples and the plots load on CircleCI now. Phys2bids version aside, I think this looks great.

Screen Shot 2020-10-26 at 12 54 53 PM

**Comment by eurunuela** [comment]
Should we give users an option to decide how many points they want to plot?

**Comment by eurunuela** [comment]
I see the values are not shown in the hover tool. Did this happen to you @62442katieb ?

**Comment by 62442katieb** [comment]
> I see the values are not shown in the hover tool. Did this happen to you @62442katieb ?

Yeah, I was having the same issue. I'll take a look and see if I can't figure it out

**Comment by 62442katieb** [comment]
Unless your recent commit fixed the issue?

**Comment by eurunuela** [comment]
> Unless your recent commit fixed the issue?

No! I didn't find a way to solve it. To me the code looks fine, so I'm not sure what's going on there. I wanted to know whether the issue was already there when you first started working on this, hoping we could track it down to the time it worked (which apparently never happened!)

**Comment by 62442katieb** [comment]
> > Unless your recent commit fixed the issue?
> 
> No! I didn't find a way to solve it. To me the code looks fine, so I'm not sure what's going on there. I wanted to know whether the issue was already there when you first started working on this, hoping we could track it down to the time it worked (which apparently never happened!)

It worked very briefly when I started working on the report. I'll see if I can't fix it later today

**Comment by eurunuela** [comment]
No rush and thank you so much!

I think this is ready once we fix that.

**Comment by eurunuela** [comment]
Thank you so much @62442katieb ! I'll make the tests pass and that should be it.

**Comment by 62442katieb** [comment]
> I downloaded some of the files locally and it ended up rendering the images after a long delay. I don't think I downloaded _enough_ of the files for the report to look 100%, and it freezes every time I move a figure, but it's enough to say that this might just be a time issue.
>>We could just show just part of the time series to fix the loading issue, but I'd expect users to want to see the entire thing.

@tsalo and @eurunuela, I fixed the load time vs. time series truncation issue by downsampling to approx. 100Hz just for plotting the signals 👍

**Comment by eurunuela** [comment]
Apparently, @62442katieb 's changes were merged into my master branch and not the one linked to this PR.

I've just added those changes. Now yes, we could say the reports are ready.

**Comment by eurunuela** [comment]
Okay, if that last commit didnt' fix the Windows error, I don't know what will... any idea @62442katieb @tsalo @smoia @vinferrer ?

**Comment by smoia** [comment]
From the reports, all the issues that I see are coming from the lack of a directory (`assets`) or an issue with the specified path (?).
I don't think it's related to the package functionality. Maybe there's something in the tests?

**Comment by eurunuela** [comment]
> From the reports, all the issues that I see are coming from the lack of a directory (`assets`) or an issue with the specified path (?).

Indeed, the problem is with the `assets` folder. However, it only happens on Windows.

> I don't think it's related to the package functionality. Maybe there's something in the tests?

It's not the tests, but a line of code in `html_report.py` that copies the assets to the output directory for the correct visualization of the report (CSS, logo, etc.)

File: phys2bids/reporting/html_report.py
Comment (smoia) [reply phys2bids/reporting/html_report.py:126]: Actually, not sure this is the right place for this? Maybe it's the next function?

File: phys2bids/reporting/html_report.py
Comment (smoia) [inline phys2bids/reporting/html_report.py:187]:
```suggestion

    Notes
    -----
```

File: phys2bids/tests/test_integration.py
Comment (smoia) [inline phys2bids/tests/test_integration.py:13]: Although I like it, I'm not sure this is the best way to go - can you reorder a little the phys2bids imports?

Thinking about it, shouldn't it be:
```
import .
from ._version import get_versions
```
etc.etc.?

File: setup.cfg
Comment (smoia) [inline setup.cfg:25]: I like the html plots A LOT, but I would not make them "mandatory", rather suggested.
I might not _always_ want an html report, for instance if I batch a bunch of files in HCPs. 
In that case we can add bokeh as an extra dependency. What do you @physiopy/all think?

File: phys2bids/phys2bids.py
Comment (smoia) [inline phys2bids/phys2bids.py:440]: As I wrote elsewhere in this review, I wouldn't make this mandatory, but rather optional AND suggested. That's not because I don't like the reports (I LOVE them), but because there might be occasions where you don't want them.
In that case, we need to add a parser flag, and/or make this line run ONLY if bokeh is installed.

File: phys2bids/reporting/html_report.py
Comment (smoia) [inline phys2bids/reporting/html_report.py:4]: This is quite different from how we import `os` in any other script in the package. I can't find it now, but we changed it in the first place because it was more "standard" not to use aliases for `os`.
Could you please change it to:

```suggestion
from os.path import join
```
Or even:
```suggestion
import os
```

instead?
It's to be consistent across scripts, that's all.

File: phys2bids/reporting/html_report.py
Comment (smoia) [inline phys2bids/reporting/html_report.py:198]: Question: are we sure that a folder `assets` in the root dir of a BIDS dataset is BIDS compatible?
Maybe we should move it in `code/conversion` (or whatever we're using already), or add/append to `.gitignore` this folder.

File: phys2bids/reporting/html_report.py
Comment (smoia) [inline phys2bids/reporting/html_report.py:196]: I have the feeling that this is the line that gives us issues in Azure. Can't figure out why completely though. I have a couple of alternatives that we can evaluate

**Review by smoia — CHANGES_REQUESTED** [review]
Ok, I'm playing the role of reviewer #2 here.

Generally, it's a very good work I can't wait to see this implemented in phys2bids!

An important doubt that I have is:
** Is the `asset` folder you're creating BIDS compatible? **
If so, great. Otherwise, see my comment for two solutions. I'll try to evaluate a BIDS dataset with such folder inside and see what happens. Similarly, the reports themselves aren't BIDS compatible. We need to move them, or ask the users to specify a path for them.

Some more things:
- Check the docstrings. Most of them are scrambled together - either a copy paste went wrong or a bad merge? In any case, you need to check and correct all of them again. If that's going to slow down too much the PR, remove them - better not to have docstrings than wrong ones. But please open an issue to add them.
- Most code is quite difficult to interpret. Could you comment them some more, either inline or in the docstring?
- Check some imports, to make them coherent with the rest of the package.

Finally, about the test that is failing in Azure, I have a couple of ideas about it.
I'm quite sure the issue comes from how the path to phys2bids is expressed and interpreted by Azure. So:

**Idea 1: add the possibility to retrieve it from a remote location.**
We can put it in OSF, or here (with another PR), and `wget` the folder if needed. In that case a nice `try` statement could solve our problems - first try to copy the files, and if it's not possible, retrieve them from OSF (?).
 
**Idea 2: remove the folder "assets".**
The css file can be printed out or be placed with the other files, and the icons, well, they're nice but not necessary. Or they can be retrieved from a remote source.

**Idea 3: wait for all tests to be in CircleCI to test this one.**
However, this means figuring out a way to accept an XFail for integration tests IN WINDOWS ONLY. I'm sure there's the way to do so, but needs to be implemented.

File: phys2bids/reporting/html_report.py
Comment (eurunuela) [reply phys2bids/reporting/html_report.py]: I don't understand what you mean.

File: phys2bids/reporting/html_report.py
Comment (eurunuela) [reply phys2bids/reporting/html_report.py:205]: Yes, that's the line giving us issues in Azure. I tried different ways of solving that with no success.

File: phys2bids/tests/test_integration.py
Comment (eurunuela) [reply phys2bids/tests/test_integration.py:13]: I know it's not the nicest way of doing things but I had to because I need to access the following attribute: `p2b.__file__`.

Given that we are calling the package and the main function `phys2bids`, there's a conflict there. 

If you have a better idea of how we could deal with that, I'm all ears.

File: setup.cfg
Comment (eurunuela) [reply setup.cfg]: If it's an optional thing, we shouldn't enforce the installation. I've added it to the optional installation within the `docs` settings. I thought that made the most sense.

File: phys2bids/reporting/html_report.py
Comment (eurunuela) [reply phys2bids/reporting/html_report.py]: No idea... But it is a necessary folder. Moving to the `code/conversion` folder sounds good to me but I wonder what the best practices are in such situations.

Maybe @tsalo knows?

File: setup.cfg
Comment (62442katieb) [reply setup.cfg]: I agree with making the html report optional.

File: phys2bids/reporting/html_report.py
Comment (vinferrer) [reply phys2bids/reporting/html_report.py]: I  Guys, Nice question, you should put every non standard thing in the ``code/conversion`` folder, it is actually thought for this kind of things such as logs and reports

File: phys2bids/reporting/html_report.py
Comment (eurunuela) [reply phys2bids/reporting/html_report.py:205]: Maybe @vinferrer knows how to solve this? Could you try it on your Windows machine?

File: phys2bids/reporting/html_report.py
Comment (eurunuela) [reply phys2bids/reporting/html_report.py]: Done! Thank you guys!

**Comment by eurunuela** [comment]
I think right now the only problem is the Windows tests are not passing. I think I took care of all the comments you made on reviews.

I would be super grateful if @vinferrer could try to replicate the error as I don't have a Windows machine.

File: phys2bids/reporting/html_report.py
Comment (vinferrer) [reply phys2bids/reporting/html_report.py:205]: You mean do the report in my windows?

**Comment by vinferrer** [comment]
Okay, I am gonna try to duplicate the error, however, you folks need to update the branch to the latest in master, I suppose that's @eurunuela or @62442katieb

**Comment by vinferrer** [comment]
I can replicate the problem I think this is a blockchain thing. My best guest is that at some point the logger file is left opened and then we cannot eliminate it :
```
   for filename in glob.glob(join(conversion_path, 'phys2bids*')):
>           remove(filename)
E           PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\sento\\AppData\\Local\\Temp\\pytest-of-sento\\pytest-53\\code\\conversion\\phys2bids_2020-12-09T175335.tsv'
```

**Comment by eurunuela** [comment]
Well, that still doesn't explain why there are no issues with Linux and MacOs...

If you know what the fix should be, could you please open a PR to my branch?

**Comment by eurunuela** [comment]
Okay, I believe the problem was the fact that there was really no logger file being created in `test_integration_acq` as we're calling `phys2bids` from inside Python. 

I've moved the report test to `test_integration_heuristic` (which call phys2bids from the terminal with `subprocess`) and the issue is gone.

**Review by vinferrer — APPROVED** [review]
LGTM!!

**Comment by eurunuela** [comment]
I don't understand why one of the integration tests fails now 🤣

This PR is as crazy as 2020 lol

**Comment by eurunuela** [comment]
Okay, I think I fixed the issue now. We've got to deal with the CircleCI credits though.

**Review by vinferrer — APPROVED** [review]
If @smoia approves, we are good to go

**Comment by eurunuela** [comment]
I'd like to get 3 or 4 approvals before merging.

File: phys2bids/tests/test_integration.py
Comment (tsalo) [reply phys2bids/tests/test_integration.py:83]: Also a generic teardown might be preferable. Perhaps by using `tmp_path` or `tmp_path_factory`?

**Review by tsalo — CHANGES_REQUESTED** [review]
I have a few minor docstring and test formatting things, but I don't have any changes to request to the core code.

**Comment by eurunuela** [comment]
We have to get this merged in before ISMRM.

Any idea why the tests are failing @tsalo ? Afaik `tmp_path_factory.getbasetemp()` should already create the folder where the data is stored. I don't know what else could be raising this error. When I use a breakpoint first thing in the test, I do not see this error. I only see it when run with no breakpoints.

File: phys2bids/tests/test_integration.py
Comment (vinferrer) [inline phys2bids/tests/test_integration.py:206]: I think here we the problem,  you are deleting the testpath in the first integration test, and therefore there is an error later when trying to download to that directory

Comment (vinferrer) [inline phys2bids/tests/test_integration.py:206]:
```suggestion
```
Just delete this line

File: phys2bids/tests/test_integration.py
Comment (vinferrer) [inline phys2bids/tests/test_integration.py:79]:
```suggestion
```
Andt this one

**Comment by eurunuela** [comment]
Ok, it looks like the tests pass again. Good catch @vinferrer !

@62442katieb @smoia @vinferrer @tsalo @RayStick the PR is ready for review now!

**Comment by eurunuela** [comment]
Looks like we figured out what was going on with the logger and the Windows tests. Thank you @vinferrer for your help. 

Apparently, the log file generated by `test_integration_acq` was being read by `test_integration_heuristics` because it was the first `tsv` file in the directory. When `phys2bids` is run through Python and not the terminal, no logs are generated on Linux but apparently, an empty log file is generated on Windows. This empty log file was being read by the test, which caused the errors.

File: phys2bids/tests/test_integration.py
Comment (smoia) [reply phys2bids/tests/test_integration.py:10]: `from phys2bids import __file__`?

Also please move from pkg_resources before phys2bids (being third party non-local)

**Pull request closed without merging**
